### PR TITLE
Initial pass at supporting multiple infrastructure backends:

### PR DESCRIPTION
**Pull request by paulczar**
- depreciate --vagrant in favor of --provisioner=vagrant
- enable --provisioner=heat
  - expects ENV/heat_stack.yml
  - will also ead ENV/vars_heat.yml for any heat specific ansible var overrides.
  - expects you to source .stackrc before running for OS creds etc

### CODE REVIEW

**Comment by paulczar** [comment]
```
ursula --ursula-debug --provisioner=heat envs/example/registry site.yml --tags=common
DEBUG: Logging into heat

DEBUG: Checking for existence of heat stack: registry
DEBUG: Creating stack
DEBUG: Waiting on stack creation...
DEBUG: Waiting on stack creation...
DEBUG: Stack created!
DEBUG: server : registry01 (173.247.105.6) 
DEBUG: writing ssh_key to /tmp/ssh_key
DEBUG: sleeping until SSH is ready...
DEBUG: Running command: ansible-playbook --inventory-file envs/example/registry/hosts --user ubuntu --module-path ./library site.yml --sudo --tags=common --extra-vars @envs/example/registry/vars_heat.yml with environment: <SNIP>

PLAY [common items] ***********************************************************
Tuesday 01 September 2015  15:49:10 -0500 (0:00:02.276)       0:00:02.276 *****
===============================================================================

GATHERING FACTS ***************************************************************
Tuesday 01 September 2015  15:49:10 -0500 (0:00:00.000)       0:00:02.276 *****
ok: [registry01]

TASK: [apt-repos | add any dependent repository keys from url] ****************
Tuesday 01 September 2015  15:49:18 -0500 (0:00:08.550)       0:00:10.827 *****
skipping: [registry01] => (item={'repo': u'deb http://repo.openstack.blueboxgrid.com/bbc/ precise main', 'key_url': u'http://repo.openstack.blueboxgrid.com/blue_box_cloud.gpg.key'})
```

File: ursula_cli/shell.py [hunk 1]
```diff
@@ -223,18 +328,40 @@ def run(args, extra_args):
...
+            extra_args += ['--extra-vars', '@%s' % heat_extra_args]
+        rc = _run_heat(args=args, hot=hot)
+        if rc:
+            return rc
+        if not args.ursula_user:
+            args.ursula_user = "ubuntu"
+        if not args.ursula_sudo:
+            args.ursula_sudo = True
+        LOG.debug("sleeping until SSH is ready...")
+        time.sleep(60)
```
Comment (omgjlk) [inline ursula_cli/shell.py]: so it's just an assumption that 60 seconds is enough for ssh?

File: ursula_cli/shell.py [hunk 2]
```diff
@@ -251,18 +378,23 @@ def main():
                         help='Test syntax for playbook')
     parser.add_argument('--ursula-debug', action='store_true',
                         help='Run this tool in debug mode')
+    parser.add_argument('--provisioner',
+                        help='The external provisioner to use vagrant,heat)',
+                        default=None)
```
Comment (omgjlk) [inline ursula_cli/shell.py]: Should this have a list of supported choices? that's a thing argparse supports.

**Comment by omgjlk** [comment]
This is all cool. I think I want to make some heat modules for Ansible so we can lose all the heat code here in the cli and just call Ansible to do the heat things, and get all the good "wait for" support and such, but that's a future item. Just a couple enhancement questions inside is all.

**Comment by paulczar** [comment]
Yeah, an idea from here would be to have an `ansible` as well provisioner which would be a playbook that would stand up the infrastructure before calling the actual site.yml etc.

File: ursula_cli/shell.py (hunk 2 above)
Comment (paulczar) [inline ursula_cli/shell.py]: yeah, it should ...  this was easier to remember how to do on a plane with bad wifi.

File: ursula_cli/shell.py (hunk 1 above)
Comment (paulczar) [inline ursula_cli/shell.py]: yes ....  this is MVP and could be better with stuff like this.

**Comment by omgjlk** [comment]
Well I'm okay with merging as is, or if you want to hit a couple of my comments now instead of later that's cool too. Your choice. If you're happy with the MVP as is, feel free to merge.  :+1:

File: ursula_cli/shell.py [hunk 3]
```diff
@@ -143,6 +151,122 @@ def _vagrant_ssh_config(environment, boxes):
 
     return 0
 
+def _run_heat(args, hot):
+    from heatclient.client import Client as Heat_Client
+    from keystoneclient.v2_0 import Client as Keystone_Client
+
```
Comment (craigtracey) [inline ursula_cli/shell.py:156]: spaces

File: ursula_cli/shell.py [hunk 4]
```diff
@@ -143,6 +151,122 @@ def _vagrant_ssh_config(environment, boxes):
...
+
+    CREDS = {
+        'username': os.environ['OS_USERNAME'],
+        'password': os.environ['OS_PASSWORD'],
+        'tenant_name': os.environ['OS_TENANT_NAME'],
+        'auth_url': os.environ['OS_AUTH_URL'],
+    }
+
```
Comment (craigtracey) [inline ursula_cli/shell.py:171]: more spaces

File: ursula_cli/shell.py [hunk 5]
```diff
@@ -143,6 +151,122 @@ def _vagrant_ssh_config(environment, boxes):
...
+Host *
+  User ubuntu
+  ForwardAgent yes
+  UserKnownHostsFile /dev/null
+  StrictHostKeyChecking no
+  PasswordAuthentication no
+  IdentityFile tmp/ssh_key
+    """
+    with open("tmp/ssh_config", "w") as text_file:
+            text_file.write(ssh_config)
```
Comment (craigtracey) [inline ursula_cli/shell.py]: indentation

File: ursula_cli/shell.py [hunk 6]
```diff
@@ -223,18 +347,38 @@ def run(args, extra_args):
...
-        rc = _run_ansible(inventory, args.playbook, extra_args=extra_args, user='vagrant', sudo=True)
-        return rc
+        if not args.ursula_user:
+            args.ursula_user = "vagrant"
+        if not args.ursula_sudo:
+            args.ursula_sudo = True
+    if args.provisioner == "heat":
+        heat_file = "%s/heat_stack.yml" % args.environment
+        if not os.path.exists(heat_file):
+            raise "heat provider requires a heat file at %s" % heat_file
```
Comment (craigtracey) [inline ursula_cli/shell.py:375]: Raise an Exception("Heat...")

File: ursula_cli/shell.py [hunk 7]
```diff
@@ -243,7 +387,7 @@ def main():
     parser.add_argument('playbook', help='The playbook to run')
     # any args should be namespaced --ursula-$SOMETHING so as not to conflict
     # with ansible-playbook's command line parameters
-    parser.add_argument('--ursula-user', help='The user to run as', default='root')
+    parser.add_argument('--ursula-user', help='The user to run as', default=None)
```
Comment (craigtracey) [inline ursula_cli/shell.py]: just remove default=

**Comment by craigtracey** [comment]
just run it through pep8/autopep8

**Pull request merged**